pkg/unikernel/libvirt/vmtools.go

```diff
@@ -0,0 +1,388 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package libvirt
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"strconv"
+	"time"
+
+	"github.com/golang/glog"
+	libvirtgo "github.com/libvirt/libvirt-go"
+	libvirtxml "github.com/libvirt/libvirt-go-xml"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/frakti/pkg/unikernel/metadata"
+)
+
+const (
+	DefaultMaxMem  = 32768 // size in MiB
+	DefaultMaxCPUs = 8
+)
+
+type VMTool struct {
+	conn *LibvirtConnect
+}
+
+func NewVMTool(conn *LibvirtConnect) *VMTool {
+	return &VMTool{
+		conn: conn,
+	}
+}
+
+type VMInfo struct {
+	UUID  string
+	State libvirtgo.DomainState
+}
+
+type VMSetting struct {
+	domainName string
+	domainUUID string
+	enableKVM  bool
+	vcpuNum    int
+	memory     int
+	image      string
+}
+
+// NOTE(Crazykev): This method may be changed when support multiple container per Pod.
+// CreateContainer creates VM which contains container defined in container spec
+func (vt *VMTool) CreateContainer(ctrMeta *metadata.ContainerMetadata, sbMeta *metadata.SandboxMetadata) error {
+	settings := VMSetting{
+		domainName: sbMeta.Name,
+		domainUUID: sbMeta.ID,
+		vcpuNum:    int(sbMeta.VMConfig.CPUNum),
+		memory:     int(sbMeta.VMConfig.Memory),
+		image:      ctrMeta.ImageRef,
+		enableKVM:  enableKVM(),
+	}
+	domainxml, err := vt.createDomain(&settings)
+	if err != nil {
+		return fmt.Errorf("failed to create domain with config(%v): %v", settings, err)
+	}
+
+	if _, err = vt.conn.DefineDomain(domainxml); err != nil {
+		return err
+	}
+
+	domain, err := vt.conn.LookupDomainByUUIDString(settings.domainUUID)
+	if err != nil {
+		if domain != nil {
+			if err1 := domain.Undefine(); err1 != nil {
+				glog.Errorf("Failed to undefine failed domain: %v", err1)
+			}
+		}
+		return err
+	}
+
+	return nil
+}
+
+// StartVM starts VM by domain UUID
+func (vt *VMTool) StartVM(domainID string) error {
+	// Get domain
+	domain, err := vt.conn.LookupDomainByUUIDString(domainID)
+	if err != nil {
+		return fmt.Errorf("failed to look up domain(%q): %v", domainID, err)
+	}
+
+	// Validate domain state
+	state, _, err := domain.GetState()
+	if err != nil {
+		return fmt.Errorf("failed to get domain(%q) state: %v", domainID, err)
+	}
+	if state != libvirtgo.DOMAIN_SHUTOFF {
+		return fmt.Errorf("unexpected domain(%q) state(%v) when try to StartVM", domainID, state)
+	}
+
+	// Create domain
+	if err = domain.Create(); err != nil {
+		return fmt.Errorf("failed to create domain(%q): %v", domainID, err)
+	}
+
+	// Check domain state
+	err = wait.PollImmediate(200*time.Millisecond, 10*time.Second, domainRunning(domainID, domain))
```

Comment:
Is 10 seconds too short for nested virtualization (e.g. in dev)?

Reply:
I hope there is no 10s start time unikernel...Do you think 30s enough here?